src/language_negotiation_lambda.js

```diff
@@ -24,32 +37,98 @@ exports.handler = (event, context, callback) => {
                 sanitizedWeight = parseFloat(weight.substring(2)) || 0.0;
             }
 
-            if (supportedLanguages.includes(sanitizedLocale)){
-                selectableLanguages.push({
-                    'locale': sanitizedLocale,
-                    'weight': sanitizedWeight
-                });
+            sanitizedLanguages.push({
+                'locale': sanitizedLocale,
+                'weight': sanitizedWeight
+            });
+        });
+
+        return sanitizedLanguages;
+    }
+
+    const getSelectableLanguages = function(languages){
+        var selectableLanguages = [];
+        languages.forEach(function(language){
+            if (supportedLanguages.indexOf(language.locale) >= 0){
+                selectableLanguages.push(language);
             }
         });
 
         return selectableLanguages;
     }
 
-    try {
-        if (headers[acceptLanguage]) {
-            var selectableLanguages = getSelectableLangugaes(headers[acceptLanguage].split(','));
-            var selectedLanguage = defaultLanguage;
-            if(selectableLanguages.length > 0) {
-                selectableLanguages.sort(function(a, b){
-                    return a.weight - b.weight; // Sort Ascending
-                });
-                selectedLanguage = selectableLanguages.pop().locale;
+    const getCookieLanguage = function(cookies_str) {
+        var languageCookieName,
+            languageCookieValue,
+            selectableLanguages,
+            re;
+
+        languageCookieName = process.env.LANGUAGE_COOKIE_NAME;
+        re = new RegExp("(?:(?:^|.*;\\s*)" + languageCookieName + "\\s*\\=\\s*([^;]*).*$)|^.*$");
+
+        languageCookieValue = cookies_str.replace(re, "$1");
+
+        selectableLanguages = getSelectableLanguages(
+            [ { locale: sanitizeLocale(languageCookieValue) } ]
+        );
+
+        return getLocale(selectableLanguages);
+    }
+
+    const getBrowserLanguage = function(language_str) {
+        var languages,
+            selectableLanguages;
+
+        languages = parseAcceptLanguage(language_str);
+        selectableLanguages = getSelectableLanguages(languages);
+        selectableLanguages.sort(function(a, b){
+            return a.weight - b.weight; // Sort Ascending
+        });
+
+        return getLocale(selectableLanguages);
+    }
+
+    const run = function() {
```

Comment:
@bderusha Rather than having the try catches in the run function, would it be cleaner to have have the `getCookieLanguage`, `getBrowserLanguage`, `getSelectableLanguages` to return "" or null on error? That way each method is self contained and you would be testing the individual methods instead of the whole.